docs: [FC-0074] centralize docs for the hooks extension framework

mariajgrimaldi opened the pull request:
## Description

This PR adds documentation about the Hooks Extension Framework under the developers section to centralize common concepts between Open edX Events and filters, such as framework overview, differences, comparisons, etc.

You can see the docs rendered here: https://docsopenedxorg--599.org.readthedocs.build/en/599/developers/concepts/index.html

The documentation in https://github.com/openedx/edx-platform/tree/master/docs/hooks should be removed after this gets merged. What this docs is missing from edx-platform guides, it's the index of filters and events within edx-platform:

https://github.com/openedx/edx-platform/blob/master/docs/hooks/filters.rst#index-of-filters
https://github.com/openedx/edx-platform/blob/master/docs/hooks/events.rst#index-of-events

This PR will not replace those indexes since they need extra effort to keep up to date, making the contribution process more cumbersome since we will need to update them here each time a new event/filter is added to a service. I prefer enhancing the current list of events and filters in each repo to reference in some way where they're being used instead.

openedx-webhooks commented:
Thanks for the pull request, @mariajgrimaldi!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.


#### :radio_button: Update the status of your PR

Your PR is currently marked as a draft.  After completing the steps above, update its status by clicking "Ready for Review", or removing "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->


#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/docs-openedx-org-maintainers`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+=========================
+Hooks Extension Framework
+=========================
```
```suggestion
#########################
```

In docs.openedx.org, we use the following headings:

H1: `#` (4 lines in the character)
H2: `*` (3 lines)
H3: `=` (2 lines)
H4: `-`

mariajgrimaldi replied:
Thank you! Done: https://github.com/openedx/docs.openedx.org/pull/599/commits/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+=========================
+
+What is the Hooks Extension Framework?
+=======================================
```
```suggestion
*****************************************************
```

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/pull/599/commits/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+
+Based on the `open-closed principle`_, this framework aims to extend the Open edX platform in a maintainable way without modifying its core. The main goal is to leverage the existing extension capabilities provided by the plugin architecture, allowing developers to implement new features to fit customer needs while reducing the need for core modifications and minimizing maintenance efforts.
+
+Hooks are a list of places in the Open edX platform where externally defined functions can take place. These functions may alter what the user sees or experiences on the platform, while in other cases, they act as notifications. All hooks are designed to be extended through `Open edX Django plugins`_ and configurations.
```
```suggestion
Hooks are places in the Open edX platform where externally defined functions can take place. These functions may alter what the user sees or experiences on the platform, while in other cases, they act as notifications. All hooks are designed to be extended through `Open edX Django plugins`_ and configurations.
```

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+* `openedx-filters`_
+* `openedx-events`_
+
+The main goal of the framework is that developers can use it to change the platform's functionality as needed and still migrate to newer Open edX releases with little to no development effort. So, the framework is designed with stability in mind, meaning it is versioned and backward compatible as much as possible.
```
```suggestion
The framework's main goal is to empower developers to change the platform's functionality as needed while enabling them to migrate to newer Open edX releases with little to no development effort. The framework is designed with stability in mind, meaning it is versioned and backward compatible as much as possible.
```

mariajgrimaldi replied:
Thanks for the suggestion! Done: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+.. _open-closed principle: https://docs.openedx.org/projects/edx-platform/en/open-release-quince.master/concepts/extension_points.html
+
+Why adopt the Hooks Extension Framework?
+========================================
```
Use `*`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+========================================
+
+Stable and Maintainable Extensions
+----------------------------------
```
Use `=`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+The Hooks Extension Framework allows developers to extend the platform's functionality in a stable, maintainable, and decoupled way ensuring easier upgrades and long-term stability by removing the need to modify the core in an significant way.
+
+Contained Solution Implementation
+---------------------------------
```
Use `=`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+By avoiding core modifications, the framework promotes self-contained solutions, eliminating the need for custom code to coexist with core logic which lowers maintenance costs for extension developers.
+
+Leveraging the Open edX Plugin Extension Mechanism
+--------------------------------------------------
```
Use `=`

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+The framework allows developers to implement custom business logic and integrations directly in plugins. This keeps core modifications minimal, focusing maintenance and development efforts on plugins, where solutions can be built and maintained independently of the core platform.
+
+Standardization
+---------------
```
Use `=`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Both filters and events implementations implement an approach for adding additional features, such as communication between services or backend flow control. With these standards in place, it’s easy to identify when and how to use the framework as a solution, ensuring a consistent and predictable approach to extending the platform.
+
+Reduce Fork Modifications
+-------------------------
```
I don't think this section is needed, it's very close to the `Contained Solution Implementation` section. Maybe just add a sentence about forks in that section.

mariajgrimaldi replied:
Agreed. I think having it in the contained solutions section would be more impactful: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+The need to modify logic in forks is minimized, as most extensions can now be implementing using the framework, keeping forks closer to the core and easier to manage.
+
+Community Compatibility
+------------------------
```
Use `=`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+The framework allows for shorter and more agile contribution cycles. By adding standardized extension points, contributors avoid creating customer-specific logic, making development more community-friendly.
+
+Backward Compatibility
+-----------------------
```
Use `=`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Hooks are designed to be backward compatible, guaranteeing stability across releases and making it easier to upgrade without breaking existing functionality.
+
+Open edX Events and Filters
+============================
```
Use `*`

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+
+The main goal of the framework is that developers can use it to change the platform's functionality as needed and still migrate to newer Open edX releases with little to no development effort. So, the framework is designed with stability in mind, meaning it is versioned and backward compatible as much as possible.
+
+A longer description of the framework and its history can be found in `OEP 50`_.
```
```suggestion
A longer description of the framework and its history can be found in :doc:`openedx-proposals:architectural-decisions/oep-0050-hooks-extension-framework`
```

This will make it a bit more resilient of a link

mariajgrimaldi replied:
Would this work for all documents under docs.openedx.org? I'm not sure how this cross referencing works.

mariajgrimaldi replied:
Thank you for the suggestion! Now I'm working with the intersphinx mapping instead: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+
+A longer description of the framework and its history can be found in `OEP 50`_.
+
+.. _OEP 50: https://open-edx-proposals.readthedocs.io/en/latest/oep-0050-hooks-extension-framework.html
```
```suggestion
```

mariajgrimaldi replied:
Done! Now I'm working with the intersphinx mapping instead: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Stable and Maintainable Extensions
+----------------------------------
+
+The Hooks Extension Framework allows developers to extend the platform's functionality in a stable, maintainable, and decoupled way ensuring easier upgrades and long-term stability by removing the need to modify the core in an significant way.
```
```suggestion
The Hooks Extension Framework allows developers to extend the platform's functionality in a stable, maintainable, and decoupled way ensuring easier upgrades and long-term stability by removing the need to modify the core in a significant way.
```

mariajgrimaldi replied:
Thank you again! https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Standardization
+---------------
+
+Both filters and events implementations implement an approach for adding additional features, such as communication between services or backend flow control. With these standards in place, it’s easy to identify when and how to use the framework as a solution, ensuring a consistent and predictable approach to extending the platform.
```
```suggestion
The implementation of both filters and events allows for adding additional features, such as communication between services or backend flow control.
```

I don't really understand what the second sentence means (" With these standards in place, it’s easy to identify when and how to use the framework as a solution, ensuring a consistent and predictable approach to extending the platform.") - could you explain what you're trying to convey, so I can suggest a rewording? I'm particularly unsure of what "these standards" means.

mariajgrimaldi replied:
Events and filters solve two problems: 1. Communication and 2. Application behavior modifications, respectively. So, if a problem can be solved by one of the two, then developers can use events/filters. I think this gives developers a standard to follow for extensibility if the use case fits what I previously mentioned. Does this make sense?

mariajgrimaldi replied:
Events and filters are used for 1. Communication and 2. Application behavior modifications, respectively. So, if a problem can be solved by one of those two options, then the implementation would be pretty straightforward—at least the extension part of it. This would give developers a standard to follow for extensibility if the use case fits what I previously mentioned. Does this make sense?

mariajgrimaldi replied:
What do you think about using this instead?

> Filters and events provide developers with a standard for adding new features to the platform through extension mechanisms. Events primarily handle communication or synchronization between different parts of the application, while filters modify application behavior. If a problem can be addressed through one of these two options, the implementation becomes pretty straightforward. 

I'm not quite sure if "Standardization" is the best title for this, though.

mariajgrimaldi replied:
@sarina: what do you think about my latest suggestion?

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Backward Compatibility
+-----------------------
+
+Hooks are designed to be backward compatible, guaranteeing stability across releases and making it easier to upgrade without breaking existing functionality.
```
```suggestion
Hooks are designed to be backwards compatible, guaranteeing stability across releases and making it easier to upgrade without breaking existing functionality.
```

Does the framework provide an explicit guarantee to this effect?

mariajgrimaldi replied:
This is enforced by the versioning policy we have in place, e.g.: https://github.com/openedx/openedx-events/blob/main/docs/decisions/0002-events-naming-and-versioning.rst#consequences. I'm going to refer to the ADR in the document.

mariajgrimaldi replied:
^ Done: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+============================
+
+Open edX Events
+----------------
```
```suggestion
============
```

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Open edX Events
+----------------
+
+Events are Open edX-specific Django signals sent in specific places on the Open edX platform. They allow developers to listen to these signals and perform additional processing based on the event data.
```
```suggestion
Events are Open edX-specific Django signals sent in specific places on the Open edX platform. Developers write code that listens to these signals and performs additional processing based on the event data.
```

mariajgrimaldi replied:
Thank you for the suggestion! https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+
+To start using Open edX Events in your project, see the `Open edX Events`_ documentation.
+
+.. _Open edX Events: https://docs.openedx.org/projects/openedx-events/en/latest/
```
It'd be good to put this document in the intersphinx mapping (https://github.com/openedx/docs.openedx.org/tree/main/source#L67-L109) and reference it directly as a `:doc:`

mariajgrimaldi replied:
Sorry, I'm not sure where [source#L67-L109](https://github.com/openedx/docs.openedx.org/tree/main/source#L67-L109) should take me to.

mariajgrimaldi replied:
^ Done: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+
+To start using Open edX Filters in your project, see the `Open edX Filters`_ documentation.
+
+.. _Open edX Filters: https://docs.openedx.org/projects/openedx-filters/en/latest/
```
It'd be good to put this document in the intersphinx mapping (https://github.com/openedx/docs.openedx.org/tree/main/source#L67-L109) and reference it directly as a `:doc:`

mariajgrimaldi replied:
Agreed! https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+What is the Hooks Extension Framework?
+=======================================
+
+Based on the `open-closed principle`_, this framework aims to extend the Open edX platform in a maintainable way without modifying its core. The main goal is to leverage the existing extension capabilities provided by the plugin architecture, allowing developers to implement new features to fit customer needs while reducing the need for core modifications and minimizing maintenance efforts.
```
```suggestion
Based on the :doc:`edx-platform:open-release-quince.master/concepts/extension_points`, this framework aims to extend the Open edX platform in a maintainable way without modifying its core. The main goal is to leverage the existing extension capabilities provided by the plugin architecture, allowing developers to implement new features to fit customer needs while reducing the need for core modifications and minimizing maintenance efforts.
```

Is there a reason this document is under the quince page tree, and not its own concept document?

mariajgrimaldi replied:
^ Fixed: https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+.. _OEP 50: https://open-edx-proposals.readthedocs.io/en/latest/oep-0050-hooks-extension-framework.html
+.. _openedx-filters: https://github.com/openedx/openedx-filters
+.. _openedx-events: https://github.com/openedx/openedx-events
+.. _open-closed principle: https://docs.openedx.org/projects/edx-platform/en/open-release-quince.master/concepts/extension_points.html
```
```suggestion
```

mariajgrimaldi replied:
^ Done! https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+.. _Open edX Events: https://docs.openedx.org/projects/openedx-events/en/latest/
+
+Open edX Filters
+-----------------
```
```suggestion
================
```

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+|                    |                                                                        |  processing.                                                |
++--------------------+------------------------------------------------------------------------+-------------------------------------------------------------+
+| **Use cases**      |  Send an email notification when a user enrolls in a course.           |  Include additional information in an API endpoint response.|
+|                    |  an email notification.                                                |                                                             |
```
```suggestion
|                    |                                                                    |                                                             |
```

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+|                    |  within a Django application.                                          |  to the caller or raises exceptions during                  |
+|                    |                                                                        |  processing.                                                |
++--------------------+------------------------------------------------------------------------+-------------------------------------------------------------+
+| **Use cases**      |  Send an email notification when a user enrolls in a course.           |  Include additional information in an API endpoint response.|
```
Could you provide a more detailed example for a filter?

mariajgrimaldi replied:
I changed this example to "Prevent the enrollment of non-authorized users" https://github.com/openedx/docs.openedx.org/pull/599/commits/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+.. _Open edX Filters: https://docs.openedx.org/projects/openedx-filters/en/latest/
+
+Differences between Events and Filters
+---------------------------------------
```
```suggestion
=============================
```

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+----------------------------------------------
+
+When to use an Open edX Event?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
-----------------------------------------
```

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+.. _how to implement event receivers: https://docs.openedx.org/projects/openedx-events/en/latest/how-tos/using-events.html#receiving-events
+
+When to use an Open edX Filter?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
-----------------------------------------
```

mariajgrimaldi replied:
Thank you, done: https://github.com/openedx/docs.openedx.org/commit/a88af4d9f0b50091f95ef1ac2a07c7d1d7bd1872

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+When to use an Open edX Event?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use an Open edX Event when but not limited to:
```
```suggestion
A developer might use an Open edX Event in order to perform the following actions. Note that this is not an exhaustive list.
```

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+
+You can review the `Open edX Events`_ documentation for more information on `how to use events`_ in your project. This documentation includes a `list of available events`_ and `how to implement event receivers`_.
+
+.. _Open edX Events: https://docs.openedx.org/projects/openedx-events/en/latest/
```
This is the second definition of this in the document. I suggest compiling your links at the very end of the document, rather than the end of each section, to avoid duplicates.

mariajgrimaldi replied:
Thank you, that'd be better indeed. You can review the changes here: Here you can review the changes:

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+When to use an Open edX Filter?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use an Open edX Filter when but not limited to:
```
```suggestion
A developer might use an Open edX Filter in order to perform the following actions. Note that this is not an exhaustive list.```

mariajgrimaldi replied:
Thank you, you can review the changes here: https://github.com/openedx/docs.openedx.org/commit/c953566eae941823a601d68f712856d501fd5119

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,155 @@
+Use an Open edX Filter when but not limited to:
+
+- Enrich the data or parameters passed to a specific component, e.g., fetch reusable LTI configurations from external plugins.
+- Intercept and modify the input of a specific component, e.g., include "Edit" link to an HTML block if certain conditions are met.
```
I'm not sure this is the best example of modifying a component. What are some others?

mariajgrimaldi replied:
All these examples were extracted from a list of real-life use cases we've compiled: https://github.com/mariajgrimaldi/openedx-events-filters-analysis/blob/main/docs/Extensions_built_using_Hooks_Extensions_Framework.pdf. Since I'm writing the use-case list, that is just a filtered version of that document, I'll get some examples from there instead.

mariajgrimaldi replied:
You can review the changes here: https://github.com/openedx/docs.openedx.org/commit/c953566eae941823a601d68f712856d501fd5119

mariajgrimaldi commented:
@sarina, thank you for all the suggestions! I hope to finish addressing your review by tomorrow :)

sarina reviewed (approved):
This looks great! Before merging, can you add this document to the "What's New" section on the dev home? https://docsopenedxorg--599.org.readthedocs.build/en/599/developers/index.html

![image](https://github.com/user-attachments/assets/2e8f04a6-039a-4001-b20a-4c64e23f0108)

mariajgrimaldi commented:
@sarina, I left a comment here for you to review https://github.com/openedx/docs.openedx.org/pull/599#discussion_r1837927461. Thank you!

mariajgrimaldi commented:
@felipemontoya @Ian2012: Could you quickly verify that everything is technically correct from a framework maintainer's perspective? Thanks!

sarina commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,136 @@
+- Enforce specific constraints or business rules of a specific component, e.g., don't allow registration for non-authorized email domains.
+- Implement additional features or behavior in a specific component, e.g., add registration extra fields to the user registration form.
+
+.. For more detailed use cases, see the TODO: ADD REFERENCE TO REAL LIFE USE CASES documentation.
```
This is good to merge I guess after the real life examples docs are merged!

sarina replied:
@mariajgrimaldi before I merge, will you plan to address this in a different PR?

mariajgrimaldi replied:
I was going to, but I can quickly do it right now. I'll let you know once it's done.

felipemontoya reviewed (approved):
From the technical standpoint this PR is correct and looks good.

I have a minor concer about the reference to customers as the driving force behind developers. I don't think we should include that assumption even if right now it looks familiar for companies such as edunext, opencraft or RG.

mariajgrimaldi commented:
@felipemontoya: thank you for the review! Can you explain your concern a bit further? I don't quite understand what you mean.

mariajgrimaldi commented on source/developers/concepts/hooks_extension_framework.rst:
```diff
@@ -0,0 +1,136 @@
+What is the Hooks Extension Framework?
+**************************************
+
+Based on the :doc:`edx-platform:concepts/extension_points`, this framework aims to extend the Open edX platform in a maintainable way without modifying its core. The main goal is to leverage the existing extension capabilities provided by the `Open edX Django plugins`_ architecture, allowing developers to implement new features to fit customer needs while reducing the need for core modifications  ...
```
```suggestion
Based on the :doc:`edx-platform:concepts/extension_points`, this framework aims to extend the Open edX platform in a maintainable way without modifying its core. The main goal is to leverage the existing extension capabilities provided by the `Open edX Django plugins`_ architecture, allowing developers to implement new features while reducing the need for core modifications and minimizing maintenance efforts.
```

mariajgrimaldi replied:
@felipemontoya: would this be enough to address your concern?

mariajgrimaldi replied:
Done: https://github.com/openedx/docs.openedx.org/pull/599/commits/970122ddbb7c23d4c2f28ef9ee7a42e950d5e510

mariajgrimaldi commented:
@felipemontoya: I believe I've addressed your comment with [this commit](https://github.com/openedx/docs.openedx.org/pull/599/commits/970122ddbb7c23d4c2f28ef9ee7a42e950d5e510), thank you!

mariajgrimaldi commented:
@sarina: thank you so much! Merge when you see fit :)

sarina commented:
Should we put this in the "What's New" section on https://docsopenedxorg--599.org.readthedocs.build/en/599/developers/index.html ? And/or in the "Concepts" card? I think this is an important document to highlight!

mariajgrimaldi commented on source/developers/index.rst:
```diff
@@ -24,6 +24,7 @@ Open edX Developers
 
       * :doc:`references/running_pr_tests`
       * :doc:`references/internal_data_formats/index`
+      * :doc: `contents/hooks_extension_framework`
```
Yes, I'm realizing I did this all wrong. I'll let you know once it's done.

mariajgrimaldi commented:
@sarina: it's finally done! Thank you so much for all the help :)

sarina commented:
🎉
